input.c

```diff
@@ -336,31 +348,40 @@ static int handle_line(void *data, const char *uri)
 	struct http_get hg;
 
 	rpd->count++;
-	rpd->rc = do_http_get(&hg, uri, 0);
+	struct connection *conn = get_connection(rpd->ip);
+	rpd->rc = do_http_get(conn, &hg, uri, 0);
 	if (rpd->rc) {
 		rpd->ip->http_code = hg.code;
 		rpd->ip->http_reason = hg.reason;
 		if (hg.fd >= 0)
-			close(hg.fd);
+			connection_close(conn);
 
 		hg.reason = NULL;
 		http_get_free(&hg);
 		return 0;
 	}
 
-	rpd->rc = setup_remote(rpd->ip, hg.headers, hg.fd);
+	rpd->rc = setup_remote(rpd->ip, hg.headers);
 	http_get_free(&hg);
 	return 1;
 }
 
-static int read_playlist(struct input_plugin *ip, int sock)
+static void set_fd(struct input_plugin_data *data, int fd)
+{
+	data->fd = fd;
+	data->conn.fd_ref = &data->fd;
```

Comment:
This suggests you assume that `input_plugin_data` will already always have a connection. Which makes keeping `fd` even more irrelevant?